subprojects/docs/src/docs/userguide/migration/upgrading_version_8.adoc

```diff
@@ -38,18 +38,36 @@ The previous step will help you identify potential problems by issuing deprecati
 [[changes_8.2]]
 == Upgrading from 8.1 and earlier
 
+=== Potential breaking changes
+
+==== Kotlin DSL scripts emit compilation warnings
+
 === Deprecations
 
-[[compile_options_generated_sources_directory]]
-==== `link:{javadocPath}/org/gradle/api/tasks/compile/CompileOptions.html[CompileOptions]` method deprecations
+[[deprecated_configuration_usage]]
+==== Using configurations incorrectly
+
+Gradle will now warn at runtime when methods of link:{javadocPath}/org/gradle/api/artifacts/Configuration.html--[Configuration] are called inconsistently with the configuration's intended usage.
+
+This change is part of a larger ongoing effort to make the intended behavior of configurations more consistent and predictable, and to unlock further speed and memory improvements.
+
+Currently, the following methods should only be called with these listed allowed usages:
+
+- `resolve()` - RESOLVABLE configurations only
+- `files(Closure)`, `files(Spec)`, `files(Dependency…)`, `fileCollection(Spec)`, `fileCollection(Closure)`, `fileCollection(Dependency…)` - RESOLVABLE configurations only
+- `getResolvedConfigurations()` - RESOLVABLE configurations only
+- `defaultDependencies(Action)` - DECLARABLE configurations only
+- `shouldResolveConsistentlyWith(Configuration)` - RESOLVABLE configurations only
+- `disableConsistentResolution()` - RESOLVABLE configurations only
+- `getDependencyConstraints()` - DECLARABLE (also permitted when declaration is allowed but deprecated)
```

Comment:
I don't understand the parenthesis. Why is this permitted for configurations which are deprecated for declaration?

Reply:
See `InvalidConfigurationResolutionIntegrationTest#"fail if a dependency constraint is declared on a configuration which can not be declared against"()`.  There is specific behavior in the `compile` configuration which is tested there.

Reply:
Actually, looking deeper, this might not be needed anymore.  I'll test removing this.

Reply:
Okay, it's definitely needed.  See the many failures here: https://builds.gradle.org/buildConfiguration/Gradle_Master_Check_Stage_PullRequestFeedback_Trigger/64526152?hideProblemsFromDependencies=false&hideTestsFromDependencies=false&expandBuildChangesSection=true&expandBuildTestsSection=true

This is used (at least) by the `xyzSourceElements` configurations, which **are** declarable against, but have **declaration against deprecated**.

I'm going to leave this for now and add it as a discussion item for 8.3